modules/renter/uploadheap.go

```diff
@@ -163,59 +167,54 @@ func (r *Renter) buildUnfinishedChunks(f *file, hosts map[string]struct{}) []*un
-			if exists && !redundantPiece {
-				newUnfinishedChunks[piece.Chunk].pieceUsage[piece.Piece] = true
-				newUnfinishedChunks[piece.Chunk].piecesCompleted++
-				delete(newUnfinishedChunks[piece.Chunk].unusedHosts, hpk.String())
-			} else if exists {
-				// This host has a piece, but it is the same piece another host
-				// has. We should still remove the host from the unusedHosts
-				// since one host having multiple pieces of a chunk might lead
-				// to unexpected issues.
-				delete(newUnfinishedChunks[piece.Chunk].unusedHosts, hpk.String())
-			}
-		}
+	// Build a map of host public keys.
+	pks := make(map[string]types.SiaPublicKey)
+	for _, pk := range f.HostPublicKeys() {
+		pks[string(pk.Key)] = pk
 	}
-	// If 'saveFile' is marked, it means we deleted some dead contracts and
-	// cleaned up the file a bit. Save the file to clean up some space on disk
-	// and prevent the same work from being repeated after the next restart.
-	//
-	// TODO / NOTE: This process isn't going to make sense anymore once we
-	// switch to chunk-based saving.
-	if saveFile {
-		err := r.saveFile(f)
+
+	// Iterate through the pieces of the file and mark which hosts are already
+	// in use for the chunk. As you delete hosts from the 'unusedHosts' map,
+	// also increment the 'piecesCompleted' value.
+	for chunkIndex := uint64(0); chunkIndex < f.NumChunks(); chunkIndex++ {
+		pieces, err := f.Pieces(chunkIndex)
 		if err != nil {
-			r.log.Println("error while saving a file after pruning some contracts from it:", err)
+			r.log.Println("failed to get pieces for building incomplete chunks")
+			return nil
+		}
+		for pieceIndex, pieceSet := range pieces {
+			for _, piece := range pieceSet {
+				// Get the contract for the piece.
+				pk, exists := pks[string(piece.HostPubKey.Key)]
+				if !exists {
+					build.Critical("Couldn't find public key in map. This should never happen")
+				}
+				contractUtility, exists := r.hostContractor.ContractUtility(pk)
+				if !exists {
+					// File contract does not seem to be part of the host anymore.
+					continue
+				}
+				if !contractUtility.GoodForRenew {
+					// We are no longer renewing with this contract, so it does not
+					// count for redundancy.
+					continue
+				}
+
+				// Mark the chunk set based on the pieces in this contract.
+				_, exists = newUnfinishedChunks[chunkIndex].unusedHosts[pk.String()]
+				redundantPiece := newUnfinishedChunks[chunkIndex].pieceUsage[pieceIndex]
+				if exists && !redundantPiece {
+					newUnfinishedChunks[chunkIndex].pieceUsage[pieceIndex] = true
+					newUnfinishedChunks[chunkIndex].piecesCompleted++
+					delete(newUnfinishedChunks[chunkIndex].unusedHosts, pk.String())
+				} else if exists {
+					// This host has a piece, but it is the same piece another host
+					// has. We should still remove the host from the unusedHosts
+					// since one host having multiple pieces of a chunk might lead
+					// to unexpected issues.
```

Comment:
Can you expand this comment to explain that if we leave the host in the set of unused hosts, it's possible that one host gets multiple pieces, then the other redundant hosts on those pieces go offline, and then you have one host providing false redundancy?